### PR TITLE
feat: new PfS columns

### PR DESCRIPTION
**Pull request by frodehk**
ECALC-365

## Why is this pull request needed?

There’s currently two columns with PfS data in Centuries that is not reported in eCalc (LTP file). The columns are required for the RNB reporting, and thus important, but the data are currently not available in eCalc making it impossible to include it in the LTP file.

These data partly requires post-processing of the eCalc results and are often forgotten in the transfer from eCalc to Centuries.

## What does this pull request change?

Possibility for additional input to eCalc at generator set level, to get the wanted data in the LTP file:

- [x] Add optional input for genset: cable loss factor (loss from shore) . Used if genset category is POWER-FROM-SHORE
- [x] Calculate Power supply onshore (including cable loss): Power required * (1 + cable loss)
- [x] Add optional time series of maximum PfS (Max usage from shore)
- [x] Report power supply onshore and max usage from shore to LTP-report
- [x] Add tests

## Issues related to this change:
https://equinor-ecalc.atlassian.net/browse/ECALC-365?atlOrigin=eyJpIjoiOTg0YmIyNmQ3NWYwNDM4ZjliZWI0MDYwNDBlNDMxMjUiLCJwIjoiaiJ9

### CODE REVIEW

File: src/libecalc/core/consumers/generator_set.py [hunk 1]
```diff
@@ -48,6 +50,22 @@ def evaluate(
         fuel_rate = self.evaluate_fuel_rate(power_requirement, variables_map=variables_map)
         power_capacity_margin = self.evaluate_power_capacity_margin(power_requirement, variables_map=variables_map)
 
+        # Check if Power From Shore, if so do necessary calculations.
+        if ConsumerUserDefinedCategoryType.POWER_FROM_SHORE in self.data_transfer_object.user_defined_category.values():
+            # cable_loss = Expression.evaluate(
```
Comment (TeeeJay) [inline src/libecalc/core/consumers/generator_set.py:55]: Remove commented out code? To be used?

File: src/libecalc/presentation/exporter/queries.py [hunk 2]
```diff
@@ -276,11 +283,15 @@ def query(
                             fuel_consumer_result: GeneratorSetResult = installation_graph.get_energy_result(
                                 fuel_consumer.id
                             )
+                            power_result = fuel_consumer_result.power
+
+                            if self.power_from_shore_output == PowerFromShoreOutputType.POWER_SUPPLY_ONSHORE:
```
Comment (TeeeJay) [inline src/libecalc/presentation/exporter/queries.py:288]: I assume this is intended, but it is not clear how this works, ie. the same query is used for 3 different scenarios, depending on what is being asked for?

File: src/tests/libecalc/output/results/test_ltp.py [hunk 3]
```diff
@@ -390,3 +393,14 @@ def test_total_oil_loaded_old_method():
...
+
+def test_power_from_shore():
+    """Test power from shore output for LTP export."""
+    [
+        datetime(2027, 1, 1),
+        datetime(2028, 1, 1),
+    ]
+    regularity = 0.2
+
+    ltp_pfs_yaml_factory(regularity=regularity)
```
Comment (TeeeJay) [inline src/tests/libecalc/output/results/test_ltp.py:406]: no assert...what is being tested here?

Comment (frodehk) [reply src/tests/libecalc/output/results/test_ltp.py:406]: The test is not complete yet:-) Will do.

File: src/libecalc/presentation/exporter/queries.py (hunk 2 above)
Comment (frodehk) [reply src/libecalc/presentation/exporter/queries.py:288]: Yes, instead of having a new query for each extra PfS-column (all three columns are different power-outputs)

File: src/libecalc/core/result/results.py [hunk 4]
```diff
@@ -60,6 +60,8 @@ class GeneratorSetResult(GenericComponentResult):
 
     typ: Literal["genset"] = "genset"
     power_capacity_margin: TimeSeriesStreamDayRate
+    power_supply_onshore: Optional[TimeSeriesStreamDayRate]
+    max_usage_from_shore: Optional[TimeSeriesStreamDayRate]
```
Comment (jsolaas) [inline src/libecalc/core/result/results.py:64]: We should move it to LTP instead of having it in core as we haven't got the time to look into how to implement this the way we want it..in my opinion.

File: src/tests/libecalc/output/results/test_ltp.py (hunk 3 above)
Comment (frodehk) [reply src/tests/libecalc/output/results/test_ltp.py]: Test is done now

File: src/libecalc/core/consumers/generator_set.py (hunk 1 above)
Comment (frodehk) [reply src/libecalc/core/consumers/generator_set.py]: Fixed

File: src/libecalc/fixtures/cases/ltp_export/ltp_power_from_shore_yaml.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def ltp_pfs_yaml_factory(
+    regularity: float,
+    cable_loss: ExpressionType,
+    max_usage_from_shore: ExpressionType,
+    load_direct_consumer: float,
+) -> [Asset, VariablesMap]:
+    input_text = f"""
+    START: 2025-01-01
```
Comment (jsolaas) [inline src/libecalc/fixtures/cases/ltp_export/ltp_power_from_shore_yaml.py:21]: Can we add this yaml case as a fixture in conftest? Specifically in `valid_example_case_yaml_case`, as this will make sure we test that the model is valid with the yaml validation. I think this PR is missing the Yaml class update, similar to the update in dto.

Comment (frodehk) [reply src/libecalc/fixtures/cases/ltp_export/ltp_power_from_shore_yaml.py]: Have tried to add fixture in conftest. The Yaml class is updated.

File: src/tests/libecalc/presentation/yaml/yaml_models/test_pyyaml_yaml_model.py [hunk 6]
```diff
@@ -72,3 +72,6 @@ def test_valid_cases(self, valid_example_case_yaml_case):
                 ),
             }
         )
+
+    def test_valid_dto_cases(self, valid_example_case_dto_case):
+        valid_example_case_dto_case.ecalc_model.model_validate(valid_example_case_dto_case.ecalc_model)
```
Comment (jsolaas) [inline src/tests/libecalc/presentation/yaml/yaml_models/test_pyyaml_yaml_model.py:77]: The point was to test that the yaml was read and validated correctly, i.e. make sure we have the yaml classes etc.

I think you can remove the fixture created and this test. If we are going to add a test for this case we should set up a YamlCase. Which is basically what is being done in `ltp_pfs_yaml_factory` except it is converted to a DTOCase. For other cases we explicitly write both a YamlCase and DTOCase to make sure the mapping is correct between yaml and dto. If we have a YamlCase it can be added to the dict in `conftest` and it will automatically be tested and made sure it is valid.

**Review by jsolaas — APPROVED** [review]

File: src/tests/libecalc/presentation/yaml/yaml_models/test_pyyaml_yaml_model.py (hunk 6 above)
Comment (frodehk) [reply src/tests/libecalc/presentation/yaml/yaml_models/test_pyyaml_yaml_model.py]: Removed the checks in conftest for now.
